### ISSUE TITLE
Efficient computation of NCE scores for f5->f5 prediction

### ISSUE THREAD

**Issue by Philip-Bachman** [mila-iqia/spr#10]
This is efficient code for local->local prediction in the style of ST-DIM:

<pre><code>
def nce_per_location(f_x1, f_x2):
    '''
    Compute InfoNCE cost with source features in f_x1 and target features in
    f_x2. We assume one source feature vector per location per item in batch
    and one target feature vector per location per item in batch. There are
    n_batch items, n_locs locations, and n_rkhs dimensions per vector.
    -- note: we can predict x1->x2 and x2->x1 in parallel "for free"

    For the positive nce pair (f_x1[i, :, l], f_x2[i, :, l]), which comes from
    batch item i at spatial location l, we will use the target feature vectors
    f_x2[j, :, l] as negative samples, for all j != i.

    Input:
      f_x1 : (n_batch, n_rkhs, n_locs)  -- n_locs source vectors per item
      f_x2 : (n_batch, n_rkhs, n_locs)  -- n_locs target vectors per item
    Output:
      loss_nce : (n_batch, n_locs)       -- InfoNCE cost at each location
    '''
    n_batch = f_x1.size(0)
    n_rkhs = f_x1.size(1)
    n_locs = f_x1.size(2)
    # reshaping for big matrix multiply
    f_x1 = f_x1.permute(2, 0, 1)                        # (n_locs, n_batch, n_rkhs)
    f_x2 = f_x2.permute(2, 1, 0)                        # (n_locs, n_rkhs, n_batch)
    # compute dot(f_glb[i, :, l], f_lcl[j, :, l]) for all i, j, l
    # -- after matmul: raw_scores[l, i, j] = dot(f_x1[i, :, l], f_x2[j, :, l])
    raw_scores = torch.matmul(f_x1, f_x2)               # (n_locs, n_batch, n_batch)
    # We get NCE log softmax by normalizing over dim 1 or 2 of raw_scores...
    # -- normalizing over dim 1 gives scores for predicting x2->x1
    # -- normalizing over dim 2 gives scores for predicting x1->x2
    lsmax_x1_to_x2 = -F.log_softmax(raw_scores, dim=2)  # (n_locs, n_batch, n_batch)
    lsmax_x2_to_x1 = -F.log_softmax(raw_scores, dim=1)  # (n_locs, n_batch, n_batch)
    # make a mask for picking out the NCE scores for positive pairs
    pos_mask = torch.eye(n_batch, dtype=f_x1.dtype, device=f_x1.device)
    pos_mask = pos_mask.unsqueeze(dim=0)
    # use masked sums to select NCE scores for positive pairs
    loss_nce_x1_to_x2 = (nce_x1_to_x2 * pos_mask).sum(dim=2)  # (n_locs, n_batch)
    loss_nce_x2_to_x1 = (nce_x2_to_x1 * pos_mask).sum(dim=1)  # (n_locs, n_batch)
    # combine forwards/backwards prediction costs (or whatever)
    loss_nce = 0.5 * (loss_nce_x1_to_x2 + loss_nce_x2_to_x1)
    return loss_nce
</code></pre>
